Homework #4. Task #2. Волынец Арсений

vol0n opened the pull request.

vol0n committed:
Initial commit

SpirinEgor commented on config/detekt/detekt.yml:
```diff
@@ -115,7 +115,7 @@ complexity:
     ignoreStringsRegex: '$^'
   TooManyFunctions:
-    active: true
+    active: false
```
Если закрыть глаза, то проблема не исчезнет)
Детект не просто так говорит тебе, что в классе много функций

SpirinEgor commented on src/main/kotlin/homework4/task2/AVLTree.kt:
```diff
@@ -0,0 +1,256 @@
+    private var root: TreeNode<K, V>? = null
+
+    private fun height(x: TreeNode<K, V>?) = x?.height ?: 0
+    private fun bFactor(x: TreeNode<K, V>) = height(x.left) - height(x.right)
```
Зачем сокращать и без того короткие названия? balanceFactor и сразу всё понятно.

SpirinEgor commented on src/main/kotlin/homework4/task2/AVLTree.kt:
```diff
@@ -0,0 +1,256 @@
+    }
+
+    internal fun rotateLeft(a: TreeNode<K, V>): TreeNode<K, V> {
+        val b = a.right!!
```
За !! сразу -
Подобных вещей вообще не должно быть, язык тебя предупреждает о возможных проблемах, надо это использовать

SpirinEgor commented on src/main/kotlin/homework4/task2/AVLTree.kt:
```diff
@@ -0,0 +1,256 @@
+
+        b.left = a
+
+        if (a.p == null) {
+            root = b
+        } else if (a.p?.right == a) {
+            a.p?.right = b
+        } else {
+            a.p?.left = b
+        }
```
Здесь можно when красиво сделать

SpirinEgor commented on src/main/kotlin/homework4/task2/AVLTree.kt:
```diff
@@ -0,0 +1,256 @@
+    }
+
+    private fun bigRotateRight(x: TreeNode<K, V>): TreeNode<K, V> {
+        rotateLeft(x.left!!)
```
Опять !! :(

SpirinEgor commented on src/main/kotlin/homework4/task2/AVLTree.kt:
```diff
@@ -0,0 +1,256 @@
+        return curNode
+    }
+
+    @Suppress("MagicNumber")
```
А можно добавить в companion object константу и ничего скрывать не надо будет

SpirinEgor commented on src/main/kotlin/homework4/task2/AVLTree.kt:
```diff
@@ -0,0 +1,256 @@
+        return root?.containsValue(value) == true
+    }
+
+    private fun inOrderTreeWalk(f: (x: TreeNode<K, V>) -> Unit) {
```
Есть ряд рекомендаций по именованиям сущностей
1. Имена не должны быть короче 3 символов
2. Функции именуются глаголом -- что они делают
3. Переменные существительным -- что они хранят

Делать переменную `f` не очень хорошо, такое название не несёт какого-либо смысла

SpirinEgor commented on src/main/kotlin/homework4/task2/AVLTree.kt:
```diff
@@ -0,0 +1,256 @@
+        return res
+    }
+
+    override fun get(key: K): V? {
+        return find(key)?.value
+    }
```
Ты везде пишешь через =, кажется и тут уместно

SpirinEgor reviewed (changes requested):
Надо обязательно исправить `!!`. Ну и подумать, как не засорять класс лишними функциями. В этой задаче разрешается добавить игнор только override методов, так как от них не убежать.

vol0n committed:
Required changes: eliminate !!,reduce funcs number

- Remove !!, check reference and throw error with message instead
- Reduce number of funcs in AVLTree, move some of them to TreeNode
- Use when in cascade if-else
- Refactor some names

SpirinEgor commented on src/main/kotlin/homework4/task2/AVLTree.kt:
```diff
@@ -0,0 +1,200 @@
+    internal fun rotateLeft(a: TreeNode<K, V>): TreeNode<K, V> {
+        val secondNode = a.right
+        if (secondNode == null) {
+            println("rotateLeft failed: right child is null!")
```
Не очень хорошо, делать какие-то внутренние принты. Пользователь не всегда может их прочитать, а иногда наоборот, они только мешают. Надо бросать исключения.

SpirinEgor commented on src/main/kotlin/homework4/task2/AVLTree.kt:
```diff
@@ -0,0 +1,200 @@
+    }
+
+    internal fun rotateRight(rotatingNode: TreeNode<K, V>): TreeNode<K, V> {
+        val n = rotatingNode.left ?: error("rotateRight failed: right child is null!")
```
А тут ты уже ошибку нормально хендлишь. Какая-то неконсистентность в коде

SpirinEgor commented on src/main/kotlin/homework4/task2/AVLTree.kt:
```diff
@@ -0,0 +1,200 @@
+
+        secondNode.right = rotatingNode
+
+        when {
+            rotatingNode.p == null -> root = secondNode
+            rotatingNode.p?.right == rotatingNode -> rotatingNode.p?.right = secondNode
+            else -> rotatingNode.p?.left = secondNode
+        }
```
Опять же, тут when есть, а выше if...

SpirinEgor reviewed (changes requested):
Кажется нет каких-то важных тестов, например, проверка обхода, на котором у тебя основано несколько функций в дальнейшем.

vol0n committed:
Merge branch 'master' into hw4_task2

vol0n committed:
Add TreeNode tests, fix style
